### PR TITLE
use grpc for communicating with compactor for query time filtering of data requested for deletion

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
Add grpc support to compactor for getting delete requests and gen number for query time filtering.
Since these requests are internal to Loki, it would be good to use grpc instead of HTTP same as all the internal requests we do in Loki.

I have added a new config for accepting the grpc address of the compactor. I tried having just the existing config and detecting if it is a grpc server, but it was hard to do it reliably, considering the different deployment modes we support. I think it is safe to keep it the same and eventually deprecate the existing config.

**Checklist**
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated

### CODE REVIEW

**Comment by periklis** [comment]
@sandeepsukhani You are addressing this issue btw: #7595

**Review by MasslessParticle — APPROVED** [review]
This looks good. Nice work!

**Review by periklis — APPROVED** [review]
Looks also good to me

File: pkg/loki/modules.go [hunk 1]
```diff
@@ -713,7 +713,7 @@ func (t *Loki) supportIndexDeleteRequest() bool {
 func (t *Loki) compactorAddress() (string, bool, error) {
 	if t.Cfg.isModuleEnabled(All) || t.Cfg.isModuleEnabled(Read) {
 		// In single binary or read modes, this module depends on Server
-		return fmt.Sprintf("127.0.0.1:%d", t.Cfg.Server.GRPCListenPort), true, nil
+		return fmt.Sprintf("%s:%d", t.Cfg.Server.HTTPListenAddress, t.Cfg.Server.GRPCListenPort), true, nil
```
Comment (periklis) [inline pkg/loki/modules.go:716]:
```suggestion
		return fmt.Sprintf("%s:%d", t.Cfg.Server.GRPCListenAddress, t.Cfg.Server.GRPCListenPort), true, nil
```
Did you mean GRPC?

Comment (sandeepsukhani) [reply pkg/loki/modules.go:716]: yeah, sorry, my bad, messed up while merging conflicts. Thanks for noticing it and pointing it out!

**Pull request merged**